### ISSUE TITLE
Dark Mode

### ISSUE THREAD

**Issue by ShadowFM** [fomantic/Fomantic-UI#1424]
Hi!

I would really like to have a whole "dark mode" only a few items have a inverted mode but not everyone.

I would really like to have all of them as a inverted mode or a simillar way to get them all dark.

**Comment by lubber-de** [comment]
This "just" needs a completely new theme. Volunteers with a proper PR are welcome 😉

**Comment by luki1412** [comment]
I had to add a lot of custom css code to make all the ones I needed dark.

**Comment by naterexw** [comment]
Great work.  @lubber-de How to add to Fomantic documentation after PR gets approved?

**Comment by lubber-de** [comment]
@naterexw By providing another PR at the docs repo. At least to be added to the index page 

https://github.com/fomantic/Fomantic-UI-Docs/blob/master/server/documents/index.html.eco#L200-L210

**Comment by Jiab77** [comment]
@ShadowFM it worth what it worth but I've made a small project to solve this issue: https://github.com/Jiab77/dark-fomantic-ui

You can define which components you want to get the `inverted` class applied to. I'll put some screenshots soon so it will be easier to get a clear idea on how to use it.

The code is not perfect and needs some improvements but it is already functional.

> This is not for promotion, I just had the same issue and decided to create something to solve it. If it can be useful for someone else then I'm happy :)

**Comment by tonimelisma** [comment]
@lubber-de it seems like the PR has been stale for 6 months, and won't be finished. Time to remove it and add back the tag help wanted? Am I correct to assume that all that's missing is the theme? And is there existing code which will chose the theme based on a media query indicating light/dark mode support?

**Comment by lubber-de** [comment]
@tonimelisma Yes, only the theme is missing. For usage of media query for dark/light mode i suggested a simple solution at https://github.com/fomantic/Fomantic-UI/discussions/1022#discussioncomment-92727

**Comment by tonimelisma** [comment]
@lubber-de so the developer would need to link three different stylesheets to the head of the HTML page? Do you think it would be possible to just do a similar trick within one CSS file, kind of like water.css does it:

[https://watercss.kognise.dev/](https://watercss.kognise.dev/)

The CSS basically defines the light color scheme, block by block, and after each block, has another with ```@media (prefers-color-scheme: dark)``` which then overwrites the light color scheme with dark versions, if the user so wants. This is automatic for both the developer as well as the end user, and is also compliant with browsers (IE) that don't support the media query (they just end up using the light color scheme).

Would something like this fit fomantic architecturally?

**Comment by lubber-de** [comment]
@tonimelisma watercss uses the css variables approach.
Providing light+dark values (as well as using css variables instead of LESS variables) for every possible element block by block inside the same theme would mean you always have to  maintain both schemes inside one single theme. That said, we could perhaps also question/discuss the existence of `inverted` elements then.

As you can see in the above mentioned dark theme PR, it's not that easy to provide a new theme which covers all components, so it needs even more effort to implement everything inside only one theme. (Volunteers? 😉 )

However, it would a complete different design approach and at least not backward compatible for FUI/SUI 2.x. That was the reason i suggested the simple imports of separate themes per media query which is backward compatible and each theme is still maintainable on its own.

We discussed new ways of theming for v3 in #441

**Comment by tonimelisma** [comment]
Right, so perhaps "built-in support" for the media query in v3, but until then, the fomantic approach is just to have two themes which are linked separately in <head> - did I understand this correctly?

**Comment by lubber-de** [comment]
Yes... that's at least my opinion 😉

**Comment by Jiab77** [comment]
I've deleted my previous comment because nobody reacted and was probably mistaken for promotion. What I mean is that your framework already contains everything to make a cleaner dark mode. On my project related to yours and apply the dark mode theme on the fly, I'm using `@media (prefers-color-scheme: dark)` to detect the user preference on page load, then will simply apply your magic `inverted` class over elements that has to be changed.

The advantage I'm seeing with this approach is that you don't have to write any specific `CSS` file to manage elements that has to be styled dark. The current implementation is not the best but while reading comments in this issue I'll probably use a `JSON` file.

Anyway, all of this to say that your framework is really nice and with skilled guys like you, I'm sure you'll made a dark mode just perfect ;)

**Comment by Jiab77** [comment]
@lubber-de I leave my comment above but by looking at https://github.com/fomantic/Fomantic-UI/issues/441#issuecomment-458942771, this structure seems to be a good idea. it would eventually make the toggle between light and dark very simple :)

**Comment by agjohnson** [comment]
So, I ended up coming up with a mild hack here that achieves dark mode without redefining colors or using a separate theme, and which will be able to work with media query color preference switching. It relies on LESS `extends()` to override the non-inverted style rules with the inverted style rules -- ie `.ui.button` extends `.ui.inverted.button` when color preference is dark.

To give an idea of what this looks like in effect, here's my very rough initial pass:

- https://github.com/readthedocs/ext-theme/pull/727

Now, this is fiddly because the extends targets are all selectors from _the final CSS rules selectors_, not LESS source selectors. A more polished version of this would be to utilize mixins or something for the inverted styles, but that's a much larger change.

Is this too much of a hack to implement in FUI core? It could exist as something separate if needed, I could see this being a maintenance annoyance. Either way, I don't feel this should be default behavior, and perhaps I need some guidance on what patterns would work best if this is worth existing somewhere.

**Comment by agjohnson** [comment]
My idea here progressed into something that should be pretty reusable, though it's worth reiterating that this is a big ol' hack. I ended up implementing a PostCSS plugin that transforms the FUI CSS to provide the dark theme overlay CSS. This isn't released as packages, but source is here:

- https://github.com/agjohnson/postcss-fomanticui-dark

CSS only install build is at:

- https://github.com/agjohnson/postcss-fomanticui-dark/tree/main/packages/fomantic-ui-dark-css

The plugin is pretty basic, this repo is mostly test cases testing against the default theme. This will probably choke on other themes and the install directions are probably way off (because Webpack is Webpack, good luck). But I'd be into any other use cases or testing here.
